### DISCUSSION TITLE
Connectiong to K/Kore: injections

### DISCUSSION THREAD

**Discussion by h0nzZik** [harp-project/AML-Formalization#377] [Ideas]
@berpeti I am posting here a 'draft proposal' of the `inj` change that I would like to discuss with the K team.

TL;DR: In Kore, `inj` should be considered a built-in operator instead of a sort-overloaded symbol. This would simplify the semantics of Kore a lot, and most likely simplify the backends also a bit.

## Proposal 

Currently, subsorting in Kore is done by means of the sort-overloaded symbol `inj{From, To}` that has to be declared in the *.kore file, like this:
```
symbol inj{From, To}(From) : To [sortInjection{}()]
```
. From a language-design perspective, this has some drawbacks:
1. The declaration depends on the `sortInjection` attribute, which has very tricky semantics. Not only it is not clear what concrete symbols it actually defines (does it declare the symbol `inj{A,B}` if A,B are unrelated sorts?), but also the interpretation of the symbols is very hard to specify, especially in the presence of “diamond sub-sorting”. This makes precise documentation of Kore, let alone formal semantics, unnecessarily challenging.
2. It lowers the abstraction level of Kore from order-sorted to multi-sorted, by requiring that what is essentially upcasting operator is always interpreted as a matching logic symbol. (The other option, interpreting upcasting operator as an identity function on metalevel, would be simpler when documenting/formalizing Kore.)
3. It is not clearly specified what should happen in the presence of multiple symbols with the `sortInjection` attribute.
4. It requires the “user” (be it the frontend or whoever writes the test-cases for backends) to always include the same piece of Kore code  again and again.
5. It is the only situation in Kore where we need sort-parametric symbols. (I am not sure if that is really the case, please correct me if I am wrong here.)

I will explain these issues in more detail below. Now, as a remedy, it would suffice if we:
1. Didn’t emit the INJ module (containing the declaration of the `inj` symbol) in the frontend
2. Formally, in the [syntax of Kore](https://github.com/runtimeverification/haskell-backend/blob/master/docs/kore-syntax.md), we would add `upcast-pattern` and simplified `application-pattern` as follows;
```
<pattern>
  ::= <variable-pattern>
    | <matching-logic-pattern>
    | <upcast-pattern>                       //< NEW
    | <application-pattern>
    | <string-literal>

<application-pattern> ::=
    <symbol-identifier> "{" "}" "(" <patterns> ")"        //< no sorts parameters

<upcast-pattern> ::=
    "inj" "{" <sort> "," <sort> "}" "(" <pattern> ")"    //< exactly two sort parameters and one pattern
```
That way, generators of Kore terms do not even need to change, unless they rely on sort-parametric symbols other then "inj".

### The ideal, simple semantics of Kore (after applying the proposed changes)

We would have order-sorted signatures and order-sorted models. In the model, there would be a bunch of carrier sets similarly as in a many-sorted case; however, unlike in the many-sorted case, the carrier sets would be ordered by set inclusion in the same way the sorts are ordered by the subsort relation. The `inj` operator would be part of the logic and would be interpreted as identity. And there would be no `sortInjection` attribute.

### The ugly options for the semantics of the current Kore.

I can give semantics to the Kore even in the presence of the `inj-as-symbol`, however, it would be much more messy and I am afraid that nobody would like it and nobody would want to read it. So here are a few options.
1. Require that each *.kore starts with the INJ module verbatim:
```
 module INJ 
     symbol inj{From, To}(From) : To [sortInjection{}()]
     axiom {S1, S2, S3, R} \equals{S3, R}(inj{S2, S3}(inj{S1, S2}(T:S1)), inj{S1, S3}(T:S1)) [simplification{}()]
endmodule
[]
```
which has no influence on the interpretation of the rest of the file, and every other module must `import INJ`, and prohibit the use of the `sortInjection` after that. And then use the ideas from the "ideal, simple semantics of Kore" from the above section for the interpretation of the rest of the file.

2. Say that the `sortInjection` attribute is allowed only on symbol declarations, in which case it means that the symbol `inj` is not really a symbol but an upcasting operator - and proceed as in the ideal case.

3. Make the `symbol` declaration always declare a symbol, or a symbol family if it is parameterized, but if the `sortInjection` attribute is present on a symbol declaration with two sort parameters, then the symbol family is not indexed by all pairs of sorts, but only of such pairs which are in the subsorting relation. Also, make sure that the symbols that _are_ declared have the right interpretation: they are used to build something-like-new-terms, with something-like-no-junk-no-confusion properties, except that if there are multiple paths between the upper and lower sort, then some sharing has to happen. As a concrete example, consider sorts `A,B,C,D` with `A < B`, `A < C`, `B < D`, and `C < D`. That is a valid (although nonlinear) ordering. The interpretation of the symbol `inj` has to ensure that `inj{B, D}(inj{A,B}(x)) == inj{C, D}(inj{A,C}(x))` and at the same time `inj{B, D}(x) <> y`  whenever `y` is _pure_ D - that is, if `y` is not the result of casting something else to `y`. And this would have to be specified generally in the documentation of Kore, which is probably doable, but _maybe_ not really readable.

4. Make the `symbol` declaration always declare a symbol (or a symbol family) with no exceptions, and use the `sortInjection` attribute to specify only the interpretation of the declared symbols. In particular, injections outside the subsorting relation would be interpreted as bottom / empty set. However, then the backends cannot complain when the user uses seemingly incorrect injections, because they would actually be correct - just empty. And intrepretation of the "good" symbols would be as in the point (3).

In the cases (2), (3), and (4) what would happen if (a) no such declaration is present, or (b) there are two symbols with the `sortInjection` attribute?

Maybe there are some other options that I left unexplored; feel free to comment on them, too. But my main point is: we treat `inj` very specially, so maybe we should stop pretending it is just a normal symbol with some lightweight attribute on it. The `sortInjection` attribute is pretty heavy and hard to document properly. And how do we know a backend is correct if we can't even precisely describe what does it do?

Also, for the context: the precise semantics of Kore is needed for the Coq backend, which will be based on the [matching logic formalization available here](https://github.com/harp-project/AML-Formalization/).

**Comment by berpeti** [comment]
I have the following suggestions:

> . From a language-design perspective, this has some drawbacks:

Delete "."

> Not only it is not clear what concrete symbols it actually defines (does it declare the symbol inj{A,B} if A,B are unrelated sorts?), but also the interpretation of the symbols is very hard to specify, especially in the presence of “diamond sub-sorting”.

First, it is not clear what concrete symbols it actually defines (does it declare the symbol `inj{A,B}` if `A`,`B` are unrelated sorts, e.g., `nat` and `unit`?). Second, the interpretation of the symbols is very hard to specify (for example, what is the meaning of `inj{nat, unit}(5)`), especially in the presence of “diamond sub-sorting” (e.g., if `A < B,C < D` where `<` denotes subsorting, what is the relation between `inj{A, D}(a)`, `inj{B,D}(inj{A, B}(a))`, and  `inj{C,D}(inj{A, C}(a))`).

I suggest adding examples that illustrate these issues (not necessarily the ones I suggested). I see that the same example is explained later, I would move it here (or add it here too).

> Didn’t emit the INJ module (containing the declaration of the inj symbol) in the frontend

It is not clear what is currently emitted (the previous point 4 suggests only the symbol declaration is emitted).

> That way, generators of Kore terms do not even need to change, unless they rely on sort-parametric symbols other then "inj".

Are you sure about this statement?

> The ideal, simple semantics of Kore (after applying the proposed changes)

Simple semantics of Kore after applying the proposed changes.

> The ugly options for the semantics of the current Kore.

Less convenient options for the semantics of the current Kore.

> Require that each *.kore starts with the INJ module verbatim:

This is how it currently works, right? Is this a proposed change? If so, I suggest rephrasing this sentence by "We keep the INJ module at the beginning of all *.kore files". The `import INJ` and check on `sortInjection` are currently implemented or are these proposed here?

>  something-like-new-terms, with something-like-no-junk-no-confusion

These are not really well-defined terms. I would say "optional new terms with axioms similar to no-junk and no-confusion". From the text, it was also not clear for me when new terms should be constructed. Is diamond subsorting the only exception?

Also, I would still add the same example to the beginning (too), as I said above.

> And intrepretation of the "good" symbols would be as in the point (3).

"point (3)" -> "previous point", because there are multiple point (3)-s before.

> if (a) no such declaration is present

What declaration? The INJ module? It is not completely clear from the context.

> Maybe there are some other options that I left unexplored; feel free to comment on them, too. But my main point is: we treat inj very specially, so maybe we should stop pretending it is just a normal symbol with some lightweight attribute on it. The sortInjection attribute is pretty heavy and hard to document properly. And how do we know a backend is correct if we can't even precisely describe what does it do?

I suggest omitting this paragraph.
